File: Leetcode Weekly Contests/Weekly-Contest-380/Find Beautiful Indices in the Given Array I/Solution.cpp
#include <bits/stdc++.h>
using namespace std;

class Solution {
public:
    vector<int> beautifulIndices(string s, string a, string b, int k) {
        vector<int> posa; // Vector to store indices of a
        vector<int> posb; // Vector to store indices of b

        size_t pos = s.find(a); // Find the first occurrence of a

        while (pos != std::string::npos) {
            
            posa.push_back(static_cast<int>(pos));

            // Continue searching for the substring after the last found position
            pos = s.find(a, pos + 1);
        }
        
        pos = s.find(b); // Find the first occurrence of b

        while (pos != std::string::npos) {
            
            posb.push_back(static_cast<int>(pos));

            // Continue searching for the substring after the last found position
            pos = s.find(b, pos + 1);
        }
        
        int i=0, j=0;
        vector<int>ans;
        while(i < posa.size() && j < posb.size()){
            if(abs(posa[i] - posb[j]) <= k){
                ans.push_back(posa[i]);
                i++;
            }
            else{
                if(posa[i] > posb[j])
                    j++;
                else
                    i++;
            }
        }
        return ans;
    }
};

File: Leetcode Weekly Contests/Weekly-Contest-380/Find Beautiful Indices in the Given Array I/Solution.java
import java.util.*;

class Solution {
    public List<Integer> beautifulIndices(String s, String a, String b, int k) {
        List<Integer> a1 = new ArrayList<>();
        List<Integer> b1 = new ArrayList<>();
        List<Integer> ans = new ArrayList<>();
        int n = a.length();
        int m = b.length();
        
        for(int i=0; i<s.length()-n+1; i++) {
            if(s.substring(i, i+n).equals(a)) a1.add(i);
        }
        
        
        for(int i=0; i<s.length()-m+1; i++) {
            if(s.substring(i, i+m).equals(b)) b1.add(i);
        }
        
        int x=0, y = 0;
        
        while(x<a1.size() && y<b1.size()) {
            int i = a1.get(x);
            int j = b1.get(y);
            if(Math.abs(j-i) <= k) {
                ans.add(i);
                x++;
            }
            else {
                if(i<j) x++;
                else y++;
            }
        }
        
        return ans;
    }
}